iris/test.cc

```diff
@@ -0,0 +1,197 @@
+
+// Just `make` will run the test.
+// Invoke with `./.noshit/test --run=1` to spawn on `localhost:3000`.
+//
+// More info:
+// * http://en.wikipedia.org/wiki/Iris_flower_data_set
+// * http://support.sas.com/documentation/cdl/en/graphref/65389/HTML/default/images/gtdshapa.png
+// * http://www.math.uah.edu/stat/data/Fisher.html
+
+#include "../yoda/yoda.h"
+
+#include "../../Bricks/net/api/api.h"
+#include "../../Bricks/graph/gnuplot.h"
+#include "../../Bricks/strings/printf.h"
+
+#include "../../Bricks/dflags/dflags.h"
+#include "../../Bricks/3party/gtest/gtest-main-with-dflags.h"
+
+#include "iris.h"
+CEREAL_REGISTER_TYPE(LabeledFlower);
+
+using namespace bricks::strings;
+using namespace bricks::gnuplot;
+using namespace yoda;
+
+DEFINE_int32(iris_port, 3000, "");
+
+DEFINE_bool(run, false, "Set to true to run indefinitely.");
+
+// Flower ID, global and auto-increasing, for test purposes.
+size_t number_of_flowers = 0;
+std::map<size_t, std::string> dimension_names;
+
+TEST(Iris, Demo) {
+  typedef API<Dictionary<LabeledFlower>> TestAPI;
+  TestAPI api("labeled_flowers");
+
+  HTTP(FLAGS_iris_port).Register("/import", [&api](Request request) {
+    EXPECT_EQ("POST", request.method);
+    const std::string body = request.body;
+    api.Transaction([body](TestAPI::T_DATA data) {
+                      // Skip the first line with labels.
+                      bool first_line = true;
+                      for (auto flower_definition_line : Split<ByLines>(body)) {
+                        std::vector<std::string> flower_definition_fields = Split(flower_definition_line, '\t');
+                        assert(flower_definition_fields.size() == 5u);
+                        if (first_line && flower_definition_fields.back() == "Label") {
+                          // For this example, just overwrite the labels on each `/import`.
+                          dimension_names.clear();
+                          for (size_t i = 0; i < flower_definition_fields.size() - 1; ++i) {
+                            dimension_names[i] = flower_definition_fields[i];
+                          }
+                          continue;
+                        }
+                        first_line = false;
+                        // Parse flower data and add it.
+                        data.Add(LabeledFlower(++number_of_flowers,
+                                               FromString<double>(flower_definition_fields[0]),
+                                               FromString<double>(flower_definition_fields[1]),
+                                               FromString<double>(flower_definition_fields[2]),
+                                               FromString<double>(flower_definition_fields[3]),
+                                               flower_definition_fields[4]));
+                      }
+                      return Printf("Successfully imported %d flowers.\n", static_cast<int>(number_of_flowers));
```

Comment:
%zu then (
http://stackoverflow.com/questions/2125845/platform-independent-size-t-format-specifiers-in-c/22114959#22114959).
But I'm not sure Windows supports it.

Based on my prior research, casting to an int is the safest option. Another
one is to use ostringstream. :)

On Friday, May 22, 2015, Maxim Zhurovich <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> In iris/test.cc
> <https://github.com/KnowSheet/Sherlock/pull/41#discussion_r30940909>:
>
> > +                          dimension_names.clear();
> > +                          for (size_t i = 0; i < flower_definition_fields.size() - 1; ++i) {
> > +                            dimension_names[i] = flower_definition_fields[i];
> > +                          }
> > +                          continue;
> > +                        }
> > +                        first_line = false;
> > +                        // Parse flower data and add it.
> > +                        data.Add(LabeledFlower(++number_of_flowers,
> > +                                               FromString<double>(flower_definition_fields[0]),
> > +                                               FromString<double>(flower_definition_fields[1]),
> > +                                               FromString<double>(flower_definition_fields[2]),
> > +                                               FromString<double>(flower_definition_fields[3]),
> > +                                               flower_definition_fields[4]));
> > +                      }
> > +                      return Printf("Successfully imported %d flowers.\n", static_cast<int>(number_of_flowers));
>
> Why not just %lu?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KnowSheet/Sherlock/pull/41/files#r30940909>.
>


-- 
And, as always, have a nice day!
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783